### ISSUE TITLE
E0731 Error reading from socket.

### ISSUE THREAD

**Issue by MarcCoru** [nalepae/pandarallel#37]
Hi,
I tried pandarallel today and it helped speed-up my processing script quite a bit.
However, after a few minutes of processing, I get the following error every few milliseconds spammed on the screen.
No further exception is thrown and the process seems to have become stuck on this error
```
E0731 19:23:55.184354 7351 io.cc:215] Error reading from socket.
```
It is quite cryptic and I could not find anything else on this particular error. That's why I am opening this issue. I am running this on the Google Cloud Compute instance: `n1-highcpu-32` (32 vCPUs, 28.8 GB memory). I still believe that it is related to the pandarallel package since I end up with this exception when I CTRL+C out of the Error loop:

```
Traceback (most recent call last):
  File "shapesplit.py", line 290, in <module>
    main()
  File "shapesplit.py", line 58, in main
    distribute_raster_series(region, year, collection, parcels, overwrite, stop_early=stop_early)
  File "shapesplit.py", line 270, in distribute_raster_series
    distribute_raster(raster, parcels, collection, overwrite=overwrite)
  File "shapesplit.py", line 235, in distribute_raster
    result = parcels.parallel_apply(mapping_function, axis=1, args=(offset, rastername, local_filename, collection))
  File "/home/marc/miniconda3/lib/python3.7/site-packages/pandarallel/utils.py", line 74, in wrapper
    client.delete(client.list().keys())
  File "pyarrow/_plasma.pyx", line 741, in pyarrow._plasma.PlasmaClient.list
  File "pyarrow/error.pxi", line 87, in pyarrow.lib.check_status
pyarrow.lib.ArrowIOError: Connection reset by peer

Waiting up to 5 seconds.
Sent all pending logs.
Process ForkPoolWorker-15:
Traceback (most recent call last):
  File "/home/marc/miniconda3/lib/python3.7/site-packages/multiprocess/process.py", line 297, in _bootstrap
    self.run()
  File "/home/marc/miniconda3/lib/python3.7/site-packages/multiprocess/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/marc/miniconda3/lib/python3.7/site-packages/multiprocess/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/home/marc/miniconda3/lib/python3.7/site-packages/multiprocess/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/home/marc/miniconda3/lib/python3.7/site-packages/pathos/helpers/mp_helper.py", line 15, in <lambda>
    func = lambda args: f(*args)
  File "/home/marc/miniconda3/lib/python3.7/site-packages/pandarallel/dataframe.py", line 14, in worker_apply
    client = plasma.connect(plasma_store_name)
  File "pyarrow/_plasma.pyx", line 805, in pyarrow._plasma.connect
  File "pyarrow/error.pxi", line 83, in pyarrow.lib.check_status
  File "/home/marc/miniconda3/lib/python3.7/site-packages/pyarrow/compat.py", line 115, in frombytes
    def frombytes(o):
KeyboardInterrupt
```

Can you help me with that issue?

PS: here is a screenshot of the instance details
![Screenshot from 2019-07-31 21-52-24](https://user-images.githubusercontent.com/21101276/62243650-0cc72480-b3de-11e9-8b90-b5ef5995108f.png)

**Comment by MarcCoru** [comment]
I may have found an explanation. In each thread, one large image is opened and closed by one worker (with rasterio). I guess this error can happen when two workers access at the same time. It would still be nice if that case would throw a proper exception rather than get stuck on an infinite loop.

**Comment by winglight** [comment]
Same error just like yours. And I found the same code could run normally on Mac(python3.7) but not work on Ubuntu 19.10(python3.6) that report the error log as you mentioned. This ubuntu pc could run the same code with 18.10 that's upgraded to 19.10 by now.

I have no other environments to test more possiblilities. Only I can say, you should change another pc to work around it.

**Comment by nalepae** [comment]
I'm currently developing a new version of **Pandarallel** without **PyArrow Plasma** (which seems to be the cause of your bug).

This version is not yet relased, but you can already try it by:
- Cloning this git repository on your computer
- Switching to **develop** branch
- Running `$ pip install .` (with the dot after the install)

By default, this version of **Pandarallel** will try to use `/dev/shm` to transfer data between the main process and the workers. If you don't want to use this feature but you prefer to use standard multiprocessing tranfert feature (pipe), you can disable it by passing `use_memory_fs_is_available=False` in the initialize method.

See the docstring of `initialize` for more information.

Note the `shm_size_mb` parameter of `initialize` is now deprecated since **Pandarallel** don't use **PyArrow** anymore.

In you chosoe to use `/dev/shm` to transfer data between the main process and the workers and you got a memory error, you can either:
- Increase the size of this partition, thanks to [this link](https://tontr.blogspot.com/2016/07/how-to-resize-tmpfs-devshm-in-linux.html), or
- Use standard multiprocessing tranfert feature (pipe), by passing `use_memory_fs_is_available=False` in the initialize method.

Note that I did **not** re-implemented (yet) progress bar and verbosity option in this version.

To remove this develop version and retrieve the official one:
- `$ pip uninstall pandarallel`
- `$ pip install pandarallel`

Please let me know if you encounter a new bug or if it works better now.

Regards,

Manu

**Comment by winglight** [comment]
@nalepae You're awesome. It works by your instruction. :D

**Comment by nalepae** [comment]
Fixed with `pandarallel 1.4.0`

It seems your bug comes from the usage of `pyarrow plasma`.

`pandarallel 1.4.0` does not use `pyarrow plasma` any more.
